### DISCUSSION TITLE
Force synchronization of file watcher in testing

### DISCUSSION THREAD

**Discussion by vaclavHala** [microsoft/vscode-discussions#680] [Extension Development QnA] [answered]
Hi, I want to write a test like this
``` ts
test("some test scenario for my custom task provider", async () => {
    const tasksJson = vscode.Uri.file(p.resolve(testProjectRoot, '.vscode', 'tasks.json'));
    await vscode.workspace.fs.writeFile(tasksJson, Buffer.from(JSON.stringify({
                "version": "2.0.0",
                "tasks": [{ ... }]
            }, null, 2)));
    const tasks = await vscode.tasks.fetchTasks();
    const testTask = tasks.filter(task => task.name === "my_custom_task")[0];
    // now run it and check it did the expected thing
});
```
The problem is sometimes the task I wrote into `tasks.json` does not show up in the `tasks` list returned from `fetchTasks()`.
When I somehow delay execution of the fetchTasks, e.g. by placing breakpoint on that line when debugging the test and letting it sit for a few seconds, the task is always present.
The way I currently understand vscode works is it uses file watcher provided by the OS in the background which will notify it at some point the file `tasks.py` changed, but without such notification it will not try to read the file from disk inside the `fetchTasks` but rather use some in memory cached version of the file.

Is this understanding correct?
If so, is there some way to force vscode to either refresh all files now, or have it return some callback/promise which will be satisfied the next time the fs watcher updates its state?
I tried calling some commands like `workbench.action.files.revert` and `workbench.files.action.refreshFilesExplorer` to little success.
Also, I expected using `vscode.workspace.fs` rather than plain `fs` from node would completely avoid these sort of problems as by using this API vscode knows there were some changes (which I'd expect is not the case when going with the generic `fs`), so why not complete the promise it gives me back after the file I wrote is actually "seen" by other vscode APIs?

I know simple solution to the problem is polling (using some setTimeout|Interval) until the desired state occurs, but if possible I'd much prefer to avoid that where possible and rather have explicit "flush" after which I'm guaranteed to observe the latest state.

Thanks for any hints

**Comment by isidorn** [comment]
@bpasero might be able to help or to link to some docs.

**Comment by bpasero** [comment, marked as answer]
There is no method to force await all file events. You could see if listening to the file event yourself gets you there (via `createFileSystemWatcher`), but there is no real guarantee that the task system has caught up when you get the event:

https://github.com/microsoft/vscode/blob/00039b0fcf281275f6e313bacc830aca31e9bc92/src/vscode-dts/vscode.d.ts#L12402
